pkg/csi/service/node.go

```diff
@@ -674,109 +683,272 @@ func (driver *vsphereCSIDriver) NodeGetInfo(
 		log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
 		return nodeInfoResponse, nil
 	}
-	var cfg *cnsconfig.Config
-	cfgPath = os.Getenv(cnsconfig.EnvVSphereCSIConfig)
-	if cfgPath == "" {
-		cfgPath = cnsconfig.DefaultCloudConfigPath
-	}
-	cfg, err := cnsconfig.GetCnsconfig(ctx, cfgPath)
-	if err != nil {
-		if os.IsNotExist(err) {
-			log.Infof("Config file not provided to node daemonset. Assuming non-topology aware cluster.")
-			nodeInfoResponse = &csi.NodeGetInfoResponse{
-				NodeId:            nodeID,
-				MaxVolumesPerNode: maxVolumesPerNode,
+
+	var (
+		accessibleTopology map[string]string
+		err error
+	)
+	if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.ImprovedVolumeTopology) {
+		accessibleTopology, err = fetchTopologyLabelsUsingCR(ctx, nodeName, nodeID)
+	} else {
+		var cfg *cnsconfig.Config
+		cfgPath = os.Getenv(cnsconfig.EnvVSphereCSIConfig)
+		if cfgPath == "" {
+			cfgPath = cnsconfig.DefaultCloudConfigPath
+		}
+		cfg, err := cnsconfig.GetCnsconfig(ctx, cfgPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				log.Infof("Config file not provided to node daemonset. Assuming non-topology aware cluster.")
+				nodeInfoResponse = &csi.NodeGetInfoResponse{
+					NodeId:            nodeID,
+					MaxVolumesPerNode: maxVolumesPerNode,
+				}
+				log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
+				return nodeInfoResponse, nil
 			}
-			log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
-			return nodeInfoResponse, nil
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to read CNS config. Error: %v", err)
 		}
-		return nil, logger.LogNewErrorCodef(log, codes.Internal,
-			"failed to read cnsconfig. err: %v", err)
+		accessibleTopology, err = fetchTopologyLabelsUsingVCCreds(ctx, nodeID, cfg)
+	}
+	if err != nil {
+		return nil, err
 	}
-	var accessibleTopology map[string]string
+
 	topology := &csi.Topology{}
+	if len(accessibleTopology) > 0 {
+		topology.Segments = accessibleTopology
+	}
+	nodeInfoResponse = &csi.NodeGetInfoResponse{
+		NodeId:             nodeID,
+		MaxVolumesPerNode:  maxVolumesPerNode,
+		AccessibleTopology: topology,
+	}
+	log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
+	return nodeInfoResponse, nil
+}
 
-	if cfg.Labels.Zone != "" && cfg.Labels.Region != "" {
-		log.Infof("Config file provided to node daemonset with zones and regions. Assuming topology aware cluster.")
-		vcenterconfig, err := cnsvsphere.GetVirtualCenterConfig(ctx, cfg)
-		if err != nil {
+func fetchTopologyLabelsUsingCR(ctx context.Context, nodeName, nodeID string) (map[string]string, error) {
```

Comment:
This function is introducing kubernetes specific code. For other features, we hide that logic behind an interface and try to keep CSI code unaware of the internal implementation. Is there a reason we're not doing that here?

Reply:
Good point, I will look into what I can do here.

Reply:
As this needs a lot of refactoring existing code, I will take it up in a different PR as soon as this is merged.

Reply:
Will fix this as part of https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/980